tests/integration/mounts.bats

```diff
@@ -96,14 +108,101 @@ function teardown() {
 	test_ro_cgroup_mount
 }
 
-# https://github.com/opencontainers/runc/security/advisories/GHSA-m8cg-xc2p-r3fc
-# shellcheck disable=SC2031
-function test_ro_cgroup_mount() {
-	local lines status
-	# shellcheck disable=SC2016
-	update_config '.process.args |= ["sh", "-euc", "for f in `grep /sys/fs/cgroup /proc/mounts | awk \"{print \\\\$2}\"| uniq`; do test -e $f && grep -w $f /proc/mounts | tail -n1; done"]'
+function config_add_bind_mount() {
+	src="$1"
+	dst="$2"
+	bindtype="${3:-bind}"
+	idmap="${4:-}"
+
+	mappings=""
+	if [ -n "$idmap" ]; then
+		mappings='
+			"uidMappings": [{"containerID": 0, "hostID": 100000, "size": 65536}],
+			"gidMappings": [{"containerID": 0, "hostID": 100000, "size": 65536}],
+		'
+	fi
+
+	update_config '.mounts += [{
+		"source": "'"$src"'",
+		"destination": "'"$dst"'",
+		"type": "bind",
+		'"$mappings"'
+		"options": [ "'"$bindtype"'", "rprivate" ]
+	}]'
+}
+
+function test_mount_order() {
+	for flag in userns idmap; do
+		is_set=
+		if grep -q "$flag" <<<"${1:-}"; then
+			is_set=1
+		fi
+		eval "is_$flag=$is_set"
+	done
+
+	from_ctr_bind_args=("rbind")
+
+	if [ -n "$is_userns" ]; then
+		requires root
+
+		update_config '.linux.namespaces += [{"type": "user"}]
+			| .linux.uidMappings += [{"containerID": 0, "hostID": 100000, "size": 65536}]
+			| .linux.gidMappings += [{"containerID": 0, "hostID": 100000, "size": 65536}]'
+		remap_rootfs
+	fi
+
+	if [ -n "$is_idmap" ]; then
+		requires root
+		requires_kernel 5.12
+		requires_idmap_fs .
+
+		from_ctr_bind_args+=("idmap")
+	fi
+
+	mkdir -p rootfs/{mnt,final}
+	# Create a set of directories we can create a mount tree with.
+	for subdir in a/x b/y c/z; do
+		dir="bind-src/$subdir"
+		mkdir -p "$dir"
+		echo "$subdir" > "$dir/file"
+	done
+
+	rootfs="$(pwd)/rootfs"
+	rm -rf rootfs/mnt
+	mkdir rootfs/mnt
+
+	# Create a bind-mount tree.
+	config_add_bind_mount "$PWD/bind-src/a" "/mnt"
+	config_add_bind_mount "$PWD/bind-src/b" "/mnt/x"
+	config_add_bind_mount "$PWD/bind-src/c" "/mnt/x/y"
+	config_add_bind_mount "$PWD/bind-src/a" "/mnt/x/y/z"
+	# Create a recursive bind-mount that uses part of the current tree in the
+	# container.
+	config_add_bind_mount "$rootfs/mnt/x" "$rootfs/mnt/x/y/z/x" "${from_ctr_bind_args[@]}"
+	config_add_bind_mount "$rootfs/mnt/x/y" "$rootfs/mnt/x/y/z" "${from_ctr_bind_args[@]}"
```

Comment:
Do we really support and want to support the source of the bind mount in the config.json to be inside the container rootfs? Or should we fail the validation if we detect this?

I mean, end users don't know the path for the container rootfs, Kubernetes doesn't know either... I really suspect this isn't used by users and I'm unsure we want to support it.

Reply:
That is how things work with Kubernetes, but runc supports more than just Kubernetes. There is nothing in the runtime-spec that indicates this is not allowed, and I see no reason to now allow it.

Also, this test acts as a proxy for the root propagation issue, but it's easier to test this than the root propagation stuff directly.

Reply:
I see no reason to allow it and unsure it can be a headache, especially if there is no one needing that. But sure, we can support it... :)

Reply:
Which part of the runtime-spec says it's not permitted? :smile_cat:

In fact, given that relative paths to the bundle are permitted, you don't even need to know the complete path and thus is something you could reasonably do, though maybe not in Kubernetes necessarily:

> source (string, OPTIONAL) A device name, but can also be a file or directory name for bind mounts or a dummy. Path values for bind mounts are either absolute or relative to the bundle.

There are a few cases where I could imagine it being useful (wanting to create an rbind of a complicated mount tree without having to duplicate all of the mount entries again).

And also, as I mentioned in https://github.com/opencontainers/runc/pull/3985#discussion_r1414006869, these kinds of mounts already work with runc. It's just that they were broken when runc used the `nsexec.c`-based mount source fd code.

Reply:
I didn't say it wasn't permitted. But sure, if they worked, makes sense to keep that :)